examples/factory/src/lib.rs

```diff
@@ -0,0 +1,304 @@
+    pub enum ProductState {
+        /// The product has not yet been initialized by the factory.
+        Uninitialized,
+        /// The product has been initialised by the factory.
+        Initialized(Product),
+    }
+
+    /// Errors
+    #[derive(Serialize, Debug, PartialEq, Eq, Reject, SchemaType)]
+    pub enum ProductError {
+        /// Failed parsing the parameter.
+        #[from(ParseError)]
+        ParseParams,
+        /// The originator of the transaction does not match the owner of the
+        /// instance, and thus is not authorized to initialize the contract.
+        NotAuthorized,
+        /// The product was already initialized.
+        AlreadyInitialized,
+        /// [`initialize`] was not called by a factory contract.
+        NotInitializedByFactory,
+    }
+
+    /// Construct the product contract in the uninitialized state.
+    #[init(contract = "product")]
+    pub fn init(_ctx: &InitContext, _state_builder: &mut StateBuilder) -> InitResult<ProductState> {
+        Ok(ProductState::Uninitialized)
+    }
+
+    /// Initialize the product contract, supplying its index.
+    /// The following preconditions are checked:
+    ///
+    /// * The instance must not have already been initialized. Otherwise,
+    ///   [`ProductError::AlreadyInitialized`] is returned.
+    ///
+    /// * The account invoking the current transaction must be the same account
+    ///   that created the instance. Otherwise [`ProductError::NotAuthorized`]
+    ///   is returned.
+    ///
+    /// * The method must be called by another smart contract (the factory). If
+    ///   the caller is not a contract,
+    ///   [`ProductError::NotInitializedByFactory`] is returned.
+    ///
+    /// Note that beyond checking that the caller is a smart contract, this
+    /// method does not check further properties of the caller (e.g. that it is
+    /// any particular factory smart contract).
+    #[receive(contract = "product", name = "initialize", parameter = "u64", mutable)]
+    pub fn initialize(
+        ctx: &ReceiveContext,
+        host: &mut Host<ProductState>,
+    ) -> Result<(), ProductError> {
+        let state = host.state_mut();
+        // Check that the contract has not already been initialized.
+        if let ProductState::Initialized(_) = *state {
+            Err(ProductError::AlreadyInitialized)?;
+        }
+        // Check that the invoker is the owner of the contract.
+        if ctx.invoker() != ctx.owner() {
+            Err(ProductError::NotAuthorized)?;
+        }
+        // The index is supplied as a parameter by the factory.
+        let index: u64 = ctx.parameter_cursor().get()?;
+        // This endpoint should only be called by another smart contract, namely the
+        // factory, which we record in the state.
+        let factory = match ctx.sender() {
+            Address::Contract(ca) => ca,
+            _ => Err(ProductError::NotInitializedByFactory)?,
```

Comment:
If we call the error `NotInitializedByFactory` then I suggest checking if the error matches this one factory address.

Alternatively, I would call this error `NoContract`